proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -13121,6 +13121,106 @@ lemma reply_push_not_queued[wp]:
   by (wpsimp wp: hoare_drop_imp get_sched_context_wp hoare_vcg_if_lift2 hoare_vcg_all_lift
              split_del: if_split)
 
+lemma sched_context_donate_tcb_scps_of_retract[wp]:
+  "\<lbrace>\<lambda>s. heap_refs_retract (tcb_scps_of s) (sc_tcbs_of s) \<and> pred_map_eq None (tcb_scps_of s) t\<rbrace>
+   sched_context_donate scp t
+   \<lbrace>\<lambda>rv s. heap_refs_retract (tcb_scps_of s) (sc_tcbs_of s)\<rbrace>"
+  apply (wpsimp wp: set_tcb_sched_context_valid_sched_pred update_sc_tcb_valid_sched_pred get_sc_obj_ref_wp
+              simp: sched_context_donate_def obj_at_def heap_upd_def)
+  by (auto simp: vs_all_heap_simps pred_map_simps elim!: heap_refs_retractE split: if_splits)
+
+lemma reply_remove_tcb_scps_of_retract[wp]:
+  "\<lbrace>\<lambda>s. heap_refs_retract (tcb_scps_of s) (sc_tcbs_of s)\<rbrace>
+   reply_remove caller reply
+   \<lbrace>\<lambda>rv s. heap_refs_retract (tcb_scps_of s) (sc_tcbs_of s)\<rbrace>"
```

Comment:
Use invariant form

Reply:
(Maybe some of these were moved, in which case, just ignore any changes)

Reply:
These are copied, yes, but it looks like some of them are duplicated. I will remove the other copies.